### ISSUE TITLE
No 'Access-Control-Allow-Origin' header is present on the requested resource.

### ISSUE THREAD

**Issue by xyrintech** [barryvdh/laravel-cors#239]
I am getting this issue with Laravel 5.4

This is my cors.php
<?php

return [   
    'supportsCredentials' => true,
    'allowedOrigins' => ['*'],
    'allowedHeaders' => ['*'],
    'allowedMethods' => ['GET', 'POST', 'PUT',  'DELETE'],
    'exposedHeaders' => [],
    'maxAge' => 0,
    'hosts' => [],
];

**Comment by jesperbjerke** [comment]
Also having trouble with this, since updating. Running Lumen 5.5 and cors v0.9.3

**Comment by mludi** [comment]
same issue here

`Fetch API cannot load http://localhost:3000/. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'null' is therefore not allowed access. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.` is the result from my react app

I'm on Laravel 5.5.2

**Comment by GeorgeHanson** [comment]
This is probably related to the same issue I reported #240 - using dev-master as opposed to the latest tag appears to fix this.

**Comment by mludi** [comment]
Ok, ignore my issue. I forgot to add the 'Accept': 'application/json' in my frontend app but I only added the CORS-middleware to the API-Routes. So there was an error in my API but Laravel rendered from the web route and gave my (why?!) a this cors error.

**Comment by jesperbjerke** [comment]
My issue was solved by ensuring the cors global middleware was registered first, before any other global middlewares. Using dev-master broke the preflight entirely for me.

**Comment by MadnessTech** [comment]
Hola, 

Estoy intentando realizar un post para obtener el oauth2 token desde front con Vue.js 2 (localhost) a back laravel 5.4.* (servidor remoto) y me esta arrojando este error, he intentado agregar distintas cabeceras, pero persiste.  
![image](https://user-images.githubusercontent.com/26617320/31898395-746fcd68-b7ef-11e7-9f88-6624828ba613.png)

**Comment by MadnessTech** [comment]
Hello,

I am trying to make a post to get the oauth2 token from front with Vue.js 2 (localhost) to back laravel 5.4. * (Remote server) and I am throwing this error, I tried to add different headers, but it persists.

![image](https://user-images.githubusercontent.com/26617320/31898515-c12b882c-b7ef-11e7-82c9-e5ab9af97706.png)

**Comment by saspallow** [comment]
@MadnessTech me too +1

**Comment by jigarzon** [comment]
Same problem +1

**Comment by acfasj** [comment]
I am using 0.9.3, update to 0.10.0 fixed my problem
`composer require "barryvdh/laravel-cors:0.10.0"`

**Comment by jigarzon** [comment]
Its working *randomly* with POST requests. On approximately  70% of the POST request I'm getting this headers:
`
Connection:Keep-Alive
Content-Encoding:gzip
Content-Length:2897
Content-Type:text/html; charset=ISO-8859-1
Date:Wed, 13 Dec 2017 11:00:26 GMT
Keep-Alive:timeout=5, max=99
Server:Apache/2.4.25 (Debian)
Vary:Accept-Encoding
`
But there are also some requests *on the same route and with same parameters* that receive the correct headers:
`
Access-Control-Allow-Credentials:true
Access-Control-Allow-Origin:https://localhost:4000
Cache-Control:no-cache, private
Connection:Keep-Alive
Content-Encoding:gzip
Content-Length:2634
Content-Type:text/html; charset=UTF-8
Date:Wed, 13 Dec 2017 11:03:08 GMT
Keep-Alive:timeout=5, max=99
Server:Apache/2.4.25 (Debian)
Vary:Origin,Accept-Encoding
X-RateLimit-Limit:60
X-RateLimit-Remaining:59
`

**Comment by barryvdh** [comment]
Can you try with v0.10.1? Please re-open if the issue this persists.

**Comment by jigarzon** [comment]
Hi I found that in my case, the request was handled by a function that was no returning the response, it was echoing to output and after that it was no returning the typical `return response(XX)` call, so the middleware was not called

**Comment by rossjcooper** [comment]
Hi I am also getting the same issue when I respond with a 401 http status. Successful 200 requests do not give this issue.

Laravel: v5.5.28
Laravel-Cors: v0.11.0

**Comment by barryvdh** [comment]
how are you returning the 401 in your code?

**Comment by rossjcooper** [comment]
The point that it should be sending back the 401 is in the `app/Exceptions/handler.php` file, using the [jwt-auth](https://github.com/tymondesigns/jwt-auth) package:
```php
      /**
	 * Render an exception into an HTTP response.
	 *
	 * @param \Illuminate\Http\Request $request
	 * @param \Exception               $exception
	 *
	 * @return \Illuminate\Http\Response
	 */
	public function render($request, Exception $exception)
	{
		if ($exception instanceof Tymon\JWTAuth\Exceptions\TokenExpiredException) {
			return response()->json(['error' => 'Token has expired'], $exception->getStatusCode());
		} elseif ($exception instanceof Tymon\JWTAuth\Exceptions\TokenInvalidException) {
			return response()->json(['error' => 'Token is invalid'], $exception->getStatusCode());
		} elseif ($exception instanceof \Illuminate\Auth\AuthenticationException) {
			return response()->json(['error' => 'Unauthorized'], 401);
		} elseif ($exception instanceof WebsiteTokenMissingException) {
			return response()->json(['error' => 'Unauthorized'], 401);
		}

		return parent::render($request, $exception);
	}
```

**Comment by ArtemMolotov** [comment]
Same problem. POST request.
Laravel: v5.5.32
Laravel-Cors: v0.11.0

**Comment by jigarzon** [comment]
@ArtemMolotov how are you returning the 302 response? I had this issue because I was echoing the error instead of using `return response(XX)`

**Comment by ArtemMolotov** [comment]
I think, page send 302 code when `validate` return fail. In my case if email not unique, Laravel send 302 with `Location: https://smfamily.ru/marafon` (last visited page with my form).

I saw this problem many times when `allowedOrigins => ['*']` and  `allowedOrigins => ['smfamily.ru']`. With `'allowedOrigins' => ['https://smfamily.ru']` -- only when write not unique email to form.

My code:
cors.php:
```
    'supportsCredentials' => false,
    'allowedOrigins' => ['https://smfamily.ru'],
    'allowedOriginsPatterns' => [],
    'allowedHeaders' => ['*'],
    'allowedMethods' => ['*'],
    'exposedHeaders' => [],
    'maxAge' => 0,
```
web.php:
```
Route::resource('/client/lead', 'ClientController', ['only' => [
    'store',
]])->middleware(Barryvdh\Cors\HandleCors::class);
```
ClientController:
```
use App\Client;
...
    public function store(Request $request)
    {
        $request::validate([
            'email' => 'required|email|unique:clients',
        ]);

        $client = Client::create([
             ....
        ]);

        return response()->json([
            'result' => true,
            'error' => false,
            'client' => $client
        ]);
    }
```
VerifyCsrfToken:
```
    protected $except = [
        'client/lead'
    ];
```

**Comment by ArtemMolotov** [comment]
#280 #148

**Comment by ArtemMolotov** [comment]
#278

**Comment by SEMeesha** [comment]
Try this, https://www.youtube.com/watch?v=L-hYd2DObTA&t=93s

**Comment by TNote4** [comment]
`php artisan make:middleware Cors`

App\Http\Middleware\Cors.php

`
...
    public function handle($request, Closure $next)
    {
        return $next($request)
            ->header('Access-Control-Allow-Origin', '*')
            ->header('Access-Control-Allow-Methods', 'GET, POST, PUT, PATCH, DELETE, OPTIONS')
            ->header('Access-Control-Allow-Headers', 'Content-Type, Authorizations');
    }
...
`

App\Http\Kernel.php

`
    protected $middleware = [
        ...
        \App\Http\Middleware\Cors::class,
    ];
`

**Comment by KodiakSA** [comment]
Just a heads up... 
I was receiving the same error, but had nothing to do with CORS!  The error was the exact one listed above.  However, when i ran the code i eventually found I was calling a function that was throwing the server side code into an infinite loop... fixing that prevented the problem immediately

**Comment by MarkVaughn** [comment]
What fixed it for me was putting the 'cors' middleware AFTER  the  all other middle wares (e.g                    'jwt.auth')

**Comment by SvenSchoene** [comment]
I can only confirm: putting the CORS-middleware before other middlewares that might throw errors before the CORS-middleware can do its thing solved this for me.

@TNote4: That looks pretty cool and I suppose it works. Is there any reason you suggest using your own middleware instead of just doing:
```php
    protected $middleware = [
        \Barryvdh\Cors\HandleCors::class, // <-- why not just use this?
        \App\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        ...
    ];
```

**Comment by diego-lipinski-de-castro** [comment]
@SvenSchoene because it doesnt work

**Comment by braian125** [comment]
> Console:
> 
> `php artisan make:middleware Cors`
> 
> App\Http\Middleware\Cors.php:
> 
> ```
> public function handle($request, Closure $next)
> {
>     return $next($request)
>         ->header('Access-Control-Allow-Origin', '*')
>         ->header('Access-Control-Allow-Methods', 'GET, POST, PUT, PATCH, DELETE, OPTIONS')
>         ->header('Access-Control-Allow-Headers', 'Content-Type, Authorizations');
> }
> ```
> App\Http\Kernel.php:
> 
> ```
> protected $middleware = [
>     ...
>     \App\Http\Middleware\Cors::class,
> ];
> ```

This solution should work for me, but unfortunately not! I have tried different ways and I do not achieve with the solution, will the error be in my call axios?

```
let response = new Promise((res,rej)=>{
        axios({
            method: 'post',
            url: 'http://localhost:8000/api/v1/users',
            data: {'name':'foo', 'type':'bar'},
        }).then((response) => {
            res(response);
        }).catch((error) => {
            rej(error);
        });
    });

    response.then((data) => {
        console.log(data)
    }).catch((error) => {
        console.log(error)
    });
```

I get this error

Access to XMLHttpRequest at 'http://localhost:8000/api/v1/users' from origin 'http://localhost:8081' has been blocked by CORS policy: Request header field content-type is not allowed by Access-Control-Allow-Headers in preflight response.

**Comment by braian125** [comment]
> Console:
> 
> `php artisan make:middleware Cors`
> 
> App\Http\Middleware\Cors.php:
> 
> ```
> public function handle($request, Closure $next)
> {
>     return $next($request)
>         ->header('Access-Control-Allow-Origin', '*')
>         ->header('Access-Control-Allow-Methods', 'GET, POST, PUT, PATCH, DELETE, OPTIONS')
>         ->header('Access-Control-Allow-Headers', 'Content-Type, Authorizations');
> }
> ```
> App\Http\Kernel.php:
> 
> ```
> protected $middleware = [
>     ...
>     \App\Http\Middleware\Cors::class,
> ];
> ```

works for me:

```
public function handle($request, Closure $next)
    {
        header('Access-Control-Allow-Origin: *');
        header('Access-Control-Allow-Methods: GET, POST, PATCH, PUT, DELETE, OPTIONS');
        header('Access-Control-Allow-Headers: Content-Type, Accept, Authorization, X-Requested-With, Application');
        return $next($request);
    }
```
App\Http\Kernel.php:

```
protected $routeMiddleware = [
        ...
        'cors' => \App\Http\Middleware\Cors::class
    ];
```

routes/api.php

```
Route::group(['prefix' => 'v1', 'middleware' => 'cors'], function () {
        Route::get('/users/{id?}', 'UserController@index');
        Route::post('/users', 'UserController@store');
        Route::post('/users/{id}', 'UserController@update');
        Route::delete('/users/{id}', 'UserController@destroy');
    });
```

**Comment by digitlimit** [comment]
I get this error with a setup like this:
```
 'api' => [
            'throttle:60,1',
            'bindings',
            \Barryvdh\Cors\HandleCors::class,
        ],
```

However moving cors middleware to the top of api stack resolved issue
```
 'api' => [
             \Barryvdh\Cors\HandleCors::class,
            'throttle:60,1',
            'bindings'
        ],
```

**Comment by miholeus** [comment]
+1 does not work in chrome if controller returns an error

**Comment by maxkopych** [comment]
My problem was with redirect. /api/recipes/ => /api/recipes
when I changed to /api/recipes everything working fine.

**Comment by pixagraphic** [comment]
> My problem was with redirect. /api/recipes/ => /api/recipes
> when I changed to /api/recipes everything working fine.

Thanks so much, this was my problem. A trailing slash in my frontend axios request caused laravel to answer the preflight with a "301 moved permanently" because it redirected to the non-slash url. Removing the slash solved the problem.

**Comment by Stetzon** [comment]
had the same problem and figured out that it was the trailing slash on the url in the `allowed_origins` config setting:

changing  ` 'allowed_origins' => ['https://my-api.herokuapp.com/'],`
to  ` 'allowed_origins' => ['https://my-api.herokuapp.com'],`

fixed my issue.

**Comment by VivekShingala** [comment]
> Hello,
> 
> I am trying to make a post to get the oauth2 token from front with Vue.js 2 (localhost) to back laravel 5.4. * (Remote server) and I am throwing this error, I tried to add different headers, but it persists.
> 
> ![image](https://user-images.githubusercontent.com/26617320/31898515-c12b882c-b7ef-11e7-82c9-e5ab9af97706.png)

Hi @MadnessTech 

Did you solve this one?

**Comment by mohamedmalki** [comment]
I solve the problem by adding cors middleware in top of the list:

` 
 protected $middleware = [
        \Fruitcake\Cors\HandleCors::class,
        \App\Http\Middleware\TrustProxies::class,
        \App\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
];
`

**Comment by VivekShingala** [comment]
> \Fruitcake\Cors\HandleCors::class

Yes, adding this middleware at the first place worked for me as well.
\Fruitcake\Cors\HandleCors::class

**Comment by barryvdh** [comment]
I made it more clear here: https://github.com/fruitcake/laravel-cors/commit/64fd49ea210c03d9fb1e45ce47043212f0982beb
Closing now, if you have a new issue, please create a new issue.
